samples/bluetooth/hci_vs_scan_req/src/main.c

```diff
@@ -0,0 +1,312 @@
+				 BT_LE_ADV_OPT_NOTIFY_SCAN_REQ)
+#else
+#define ADV_OPTIONS             (BT_LE_ADV_OPT_SCANNABLE | BT_LE_ADV_OPT_NOTIFY_SCAN_REQ)
+#endif
+
+/* Lower the value, closer the scanner. */
+#define ADV_TX_POWER RADIO_TXPOWER_TXPOWER_Neg8dBm
+
+static const struct bt_le_adv_param parameters = {
+	.options = ADV_OPTIONS,
+	.interval_min = ADV_MIN_INTERVAL_625MS,
+	.interval_max = ADV_MAX_INTERVAL_625MS,
+};
+
+static const struct bt_data advertisement_data[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LENGTH),
+};
+
+static const struct bt_data scan_response_data[] = {
+	BT_DATA_BYTES(BT_DATA_UUID128_ALL, 0),
+};
+
+#if defined(CONFIG_BT_EXT_ADV)
+static struct bt_le_ext_adv *ext_adv;
+#endif
+
+#if defined(CONFIG_BT_PERIPHERAL) && defined(CONFIG_BT_EXT_ADV)
+static volatile bool disconnect_event_happened;
+#endif
+
+/* Allow different TX power per role/connection. It requires CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL.
+ * same as in samples/bluetooth/hci_pwr_ctrl/src
+ */
+static void set_tx_power(uint8_t handle_type, uint16_t handle, int8_t tx_pwr_lvl)
+{
+	struct bt_hci_cp_vs_write_tx_power_level *cp;
+	struct bt_hci_rp_vs_write_tx_power_level *rp;
+	struct net_buf *buf, *rsp = NULL;
+	int err;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_VS_WRITE_TX_POWER_LEVEL, sizeof(*cp));
+	if (!buf) {
+		printk("Unable to allocate command buffer\n");
+		return;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->handle = sys_cpu_to_le16(handle);
+	cp->handle_type = handle_type;
+	cp->tx_power_level = tx_pwr_lvl;
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_VS_WRITE_TX_POWER_LEVEL,
+				   buf, &rsp);
+	if (err) {
+		uint8_t reason = rsp ?
+			((struct bt_hci_rp_vs_write_tx_power_level *)
+			  rsp->data)->status : 0;
+		printk("Set Tx power err: %d reason 0x%02x\n", err, reason);
+		return;
+	}
+
+	rp = (void *)rsp->data;
+	printk("Actual Tx Power: %d\n", rp->selected_tx_power);
+
+	net_buf_unref(rsp);
+}
+
+#if defined(CONFIG_BT_CTLR_VS_SCAN_REQ_RX)
+/* BLE standard doesn't allow the scan request event with legacy advertisements.
+ * Ref: Bluetooth Core Specification v5.3, section 7.7.65.19 "LE Scan Request Received event" :
+ *      "This event shall only be generated if advertising was enabled using the
+ *       HCI_LE_Set_Extended_Advertising_Enable command."
+ * Added a Vendor Specific command to add this feature and save 5.5 kB SRAM.
+ */
+static void enable_legacy_adv_scan_request_event(bool enable)
+{
+	struct bt_hci_cp_vs_set_scan_req_reports *cp;
+	struct net_buf *buf;
+	int err;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_VS_SET_SCAN_REQ_REPORTS, sizeof(*cp));
+	if (!buf) {
+		printk("%s: Unable to allocate HCI command buffer\n", __func__);
+		return;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->enable = (uint8_t) enable;
+
+	err = bt_hci_cmd_send(BT_HCI_OP_VS_SET_SCAN_REQ_REPORTS, buf);
```

Comment:
It seems responsibility for this feature is partially implemented in the user's app and in the host: User sends an HCI command manually, and gets a callback from the host.

It needs to be either fully inside or fully outside of the host.

Speaking as a host maintainer, the less code I have to maintain, the happier I am.
You can have it outside by using [`CONFIG_BT_HCI_VS_EVT_USER`](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/Kconfig#L1103).

Reply:
Hi @jori-nordic, thanks for your feedback.
I tried using `CONFIG_BT_HCI_VS_EVT_USER` and adding a callback for HCI Vendor-Specific Events in place of the host callback I added.
I see a problem here as the APIs to get the advertiser ID seem to be host internal and not visible at application level.
I'm talking about usage of `adv_get_legacy()` that is made in `bt_hci_le_vs_scan_req_received()`, which is coded similarly to `bt_hci_le_scan_req_received()`.
(BTW I see now an improvement in its last version that I can also add.)
Do you see a smarter way or is this really a blocking point?

Reply:
I think there is one:
The only thing we use `adv` for is for getting the identity that started advertising. Since you (the application) started advertising, then you know what identity address you have. 
Correct me if I'm wrong also, but this is not extended advertising right, so we don't have multiple adv sets running.

Reply:
Indeed the application can be built both using extended (one set only) and legacy advertising, to compare performances and memory usage. At moment I'm using only legacy adv and I'm patching with local declaration of `bt_le_adv_lookup_legacy`, as it seems that I cannot include `adv.h`, but perhaps we can even not use it.
I've also (temporarily) made public not only `bt_lookup_id_addr`, but even `bt_addr_le_is_resolved` and `bt_addr_le_copy_resolved` that are all moved in `addr.h`.
As a test, I've added the new vs_scanned that is used via `CONFIG_BT_HCI_VS_EVT_USER` and returns always false by purpose to have even the previous callback to compare if they're doing the same. Yet the results are not
totally expected and I'd like to have your opinion about what I'm doing wrong.
1. Different output for the to callbacks. The logs I get are like these ones:
```
00> vs_scanned peer 2D:C6:3C:B4:A2:01 (0x04) rssi 97
00> vs_scanned (address 2D:C6:3C:B4:A2:01 (0x04) rssi 97)
00> [00:00:17.117,340] <err> bt_hci_core: subevent 0x04
00> [00:00:17.117,431] <err> bt_adv: bt_hci_le_vs_scan_req_received peer 61:2D:C6:3C:B4:A2 (random) rssi -127 adv 0x20000018
00> scanned (address 61:2D:C6:3C:B4:A2 (random) rssi -127)
```
The first two line are from the new callback. The data taken from the passed buffers seem incorrect and  the address shifted by one byte between the two functions and the RSSI seems taken from two different parts of the buffer but both wrong.

2. In the new callback vs_scanned the operations managing the address seem redundant, as both logs before and after them log the same values (unless fixing the above point makes a difference). In such case we could remove usage of the APIs made temporarily public.

Reply:
could you push the code to another branch and post a link?

Reply:
I've pushed it into this branch

Reply:
You were missing a `pull_u8` compared to the vs handler: https://github.com/jori-nordic/zephyr/blob/68659-bsim/samples/bluetooth/hci_vs_scan_req/src/main.c#L198-L201

hci_core [pulls that byte](https://github.com/jori-nordic/zephyr/blob/68659-bsim/subsys/bluetooth/host/hci_core.c#L2466-L2469) right before handing control over to the in-stack vs-handler.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/68428270/ac5614b9-e70a-4f1d-876b-780a29e49726)

Reply:
Hi @jori-nordic, thanks a lot for finding my bug!
I tired the fix and it works. The changes I'm doing now are:
1. Remove the added callback in host and use only `BT_HCI_VS_EVT_USER`.
2. In the new callback vs_scanned, I'd remove the lines starting from calling `bt_le_adv_lookup_legacy`, as they don't seem to add anything, having the same info printed in the 2nd `printk`. This would even not require to make public the four used functions. But I'd like to have your opinion.
3. I consider that the read value rssi -127 doesn't depend on this PR changes, do you agree?

Reply:
Good to hear!
1. Perfect, I'm fine with that.
2. Don't think you need them unless you care about (host-based) private address resolution. I propose you make the PR without those lines and open another one later if you find out you needed that functionality.
3. Yeah I think it was just because I removed the power setting, you can disregard that value in the screenshot.

Reply:
Now all the above mentioned changed should be pushed. There are no more changes in the host, it's everything in the sample app.